File: Palace/Network/TPPSession.m
#import "TPPAsync.h"
#import "Palace-Swift.h"

#import "TPPSession.h"

@interface TPPSession () <NSURLSessionDelegate, NSURLSessionTaskDelegate>

@property (nonatomic) NSURLSession *session;

@end

static NSUInteger const diskCacheInMegabytes = 20;
static NSUInteger const memoryCacheInMegabytes = 2;

static TPPSession *sharedSession = nil;

@implementation TPPSession

+ (instancetype)sharedSession
{
  static dispatch_once_t predicate;
  
  dispatch_once(&predicate, ^{
    sharedSession = [[self alloc] init];
    if(!sharedSession) {
      TPPLOG(@"Failed to create shared session.");
    }
  });
  
  return sharedSession;
}

#pragma mark NSObject

- (instancetype)init
{
  if(sharedSession) {
    @throw NSGenericException;
  }
  
  self = [super init];
  if(!self) return nil;
  
  NSURLSessionConfiguration *const configuration =
    [NSURLSessionConfiguration defaultSessionConfiguration];
  
  assert(configuration.URLCache);
  
  configuration.URLCache.diskCapacity = 1024 * 1024 * diskCacheInMegabytes;
  configuration.URLCache.memoryCapacity = 1024 * 1024 * memoryCacheInMegabytes;
  
  self.session = [NSURLSession sessionWithConfiguration:configuration
                                               delegate:self
                                          delegateQueue:[NSOperationQueue mainQueue]];
  
  return self;
}

#pragma mark NSURLSessionTaskDelegate

- (void)URLSession:(__attribute__((unused)) NSURLSession *)session
              task:(__attribute__((unused)) NSURLSessionTask *)task
didReceiveChallenge:(NSURLAuthenticationChallenge *const)challenge
 completionHandler:(void (^)(NSURLSessionAuthChallengeDisposition disposition,
                             NSURLCredential *credential))completionHandler
{
  TPPLOG_F(@"NSURLSessionTask: %@. Challenge Received: %@",
            task.currentRequest.URL.absoluteString,
            challenge.protectionSpace.authenticationMethod);

  TPPBasicAuth *challenger = [[TPPBasicAuth alloc] initWithCredentialsProvider:
                               TPPUserAccount.sharedAccount];
  [challenger handleChallenge:challenge completion:completionHandler];
}

#pragma mark -

- (void)uploadWithRequest:(NSURLRequest *)request completionHandler:(void (^)(NSData *data, NSURLResponse *response, NSError *error))handler
{
  [[self.session uploadTaskWithRequest:request
                              fromData:request.HTTPBody
                     completionHandler:handler] resume];
}

- (NSURLRequest*)withURL:(NSURL *const)URL
        shouldResetCache:(BOOL)shouldResetCache
       completionHandler:(void (^)(NSData *data,
                                   NSURLResponse *response,
                                   NSError *error))handler
{
  if(!handler) {
    @throw NSInvalidArgumentException;
  }

    NSURLRequest *req;
    void (^completionWrapper)(NSData * _Nullable, NSURLResponse * _Nullable, NSError * _Nullable) = ^ void (NSData * _Nullable data, NSURLResponse * _Nullable response, NSError * _Nullable error){
      if (error) {
        NSString *dataString = [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding];
        if (dataString == nil) {
          dataString = [NSString stringWithFormat:@"datalength=%lu",
                        (unsigned long)data.length];
        }
        [TPPErrorLogger logNetworkError:error
                                    code:TPPErrorCodeApiCall
                                 summary:NSStringFromClass([self class])
                                 request:req
                                response:response
                                metadata:@{
                                  @"receivedData": dataString ?: @"N/A",
                                  @"networking context": @"NYPLSession error"
                                }];
        handler(nil, response, error);
        return;
      }

      handler(data, response, nil);
    };

  if (shouldResetCache) {
    // NB: this sledgehammer approach is not ideal, and the only reason we
    // don't use `removeCachedResponseForRequest:` (which is really what we
    // should be using) is because that method has been buggy since iOS 8,
    // and it still is in iOS 13.
    [TPPNetworkExecutor.shared clearCache];
  }

  NSString *lpe = [URL lastPathComponent];
  if ([lpe isEqualToString:@"borrow"])
    req = [[TPPNetworkExecutor.shared PUT:URL useTokenIfAvailable:false completion:completionWrapper] originalRequest];
  else
    req = [[TPPNetworkExecutor.shared GET:URL cachePolicy:NSURLRequestUseProtocolCachePolicy useTokenIfAvailable:false completion:completionWrapper] originalRequest];

  return req;
}

@end

File: Palace/Network/TPPSession.swift
import Foundation

@objc class TPPSession: NSObject, URLSessionDelegate, URLSessionTaskDelegate {

  private static let diskCacheInMegabytes: Int = 15
  private static let memoryCacheInMegabytes: Int = 1

  @objc static let sharedSession: TPPSession = {
    let session = TPPSession()
    return session
  }()

  private var session: URLSession!

  private override init() {
    super.init()

    let configuration = URLSessionConfiguration.default
    assert(configuration.urlCache != nil)
    configuration.urlCache?.diskCapacity = 1024 * 1024 * TPPSession.diskCacheInMegabytes
    configuration.urlCache?.memoryCapacity = 1024 * 1024 * TPPSession.memoryCacheInMegabytes
    configuration.urlCredentialStorage = nil

    session = URLSession(
      configuration: configuration,
      delegate: self,
      delegateQueue: .main
    )
  }

  // MARK: - URLSessionTaskDelegate

  func urlSession(
    _ session: URLSession,
    task: URLSessionTask,
    didReceive challenge: URLAuthenticationChallenge,
    completionHandler: @escaping (URLSession.AuthChallengeDisposition, URLCredential?) -> Void
  ) {
    Log.log("\(task.currentRequest?.url?.absoluteString ?? ""): Challenge Received: \(challenge.protectionSpace.authenticationMethod)")

    let challenger = TPPBasicAuth(credentialsProvider: TPPUserAccount.sharedAccount())
    challenger.handleChallenge(challenge, completion: completionHandler)
  }

  // MARK: - Public methods

  @objc func upload(with request: URLRequest, completionHandler handler: ((Data?, URLResponse?, Error?) -> Void)?) {
    session.uploadTask(with: request, from: request.httpBody, completionHandler: { data, response, error in
      handler?(data, response, error)
    }).resume()
  }

  @objc func withURL(
    _ url: URL,
    shouldResetCache: Bool,
    completionHandler handler: @escaping (Data?, URLResponse?, Error?) -> Void
  ) -> URLRequest? {
    var req: URLRequest?

    let completionWrapper: (Data?, URLResponse?, Error?) -> Void = { data, response, error in
      if let error = error {
        let dataString = data.flatMap { String(data: $0, encoding: .utf8) }
          ?? data.map { "datalength=\($0.count)" }
        TPPErrorLogger.logNetworkError(
          error,
          code: .apiCall,
          summary: String(describing: type(of: self)),
          request: req as URLRequest?,
          response: response,
          metadata: [
            "receivedData": dataString ?? "N/A",
            "networking context": "NYPLSession error"
          ]
        )
        handler(nil, response, error)
        return
      }
      handler(data, response, nil)
    }

    if shouldResetCache {
      TPPNetworkExecutor.shared.clearCache()
    }

    let lastPathComponent = url.lastPathComponent
    if lastPathComponent == "borrow" {
      req = TPPNetworkExecutor.shared.PUT(url, useTokenIfAvailable: false, completion: completionWrapper)?.originalRequest
    } else {
      req = TPPNetworkExecutor.shared.GET(url, cachePolicy: .useProtocolCachePolicy, useTokenIfAvailable: false, completion: completionWrapper)?.originalRequest
    }

    return req
  }
}
